### PR TITLE
Reject on status 0x00

### PR DESCRIPTION
**Pull request by cgewecke**
Addresses requirement that truffle-contract behave appropriately with clients who only signal transaction error via the tx receipt by setting it's status property to 0x0.

+ Updates ganache dev dep to latest: `2.1.0-beta.0` (which can be configured to omit an error from the response, like Geth/Parity)
+ Adds logic to reject with the tx hash, receipt object, error message if status is 0x0
+ Adds logic to interpret meaning of status (revert, invalid opcode, OOG on internal calls).
+ Adds tests for both client behaviors.
+ Extracts provider setup into a utility to make passing options to ganache a little cleaner in `abstractions.js`

If web3 errors we reject with the message, if status is bad, we reject with the usual transaction response, minus any log data.

Also explicitly configuring ganache to run silently in the tests because it's logging everything out to the console.  

[EDITED}

### CODE REVIEW

File: contract.js [hunk 1]
```diff
@@ -171,12 +171,20 @@ var contract = (function(module) {
...
                   if (err) return reject(err);
 
+                  // Reject on transaction failures, accept otherwise
                   if (receipt != null) {
-                    return accept({
-                      tx: tx,
-                      receipt: receipt,
-                      logs: Utils.decodeLogs(C, instance, receipt.logs)
-                    });
+                    if (receipt.status === "0x00"){
```
Comment (benjamincburns) [inline contract.js]: It's actually a bug that the status is `0x00` and not `0x0` due to the RPC quantity encoding rules.

It's probably better to do something like `parseInt(receipt.status, 16) == 0`

Comment (cgewecke) [reply contract.js]: @benjamincburns Oh! Good to know, thank you.

File: test/abstractions.js [hunk 2]
```diff
@@ -204,6 +208,59 @@ describe("Abstractions", function() {
     });
   });
 
+
+  it("errors with a message when ganache appends an error to the response", function(done){
```
Comment (benjamincburns) [inline test/abstractions.js]: It feels a bit heavy-handed to me that we'd need ganache in the loop to test this code. I think it'd be easy to break the logic which this test exercises out into a separate utility function and write a very thin unit test for it.

File: test/cloning.js [hunk 3]
```diff
@@ -32,8 +34,8 @@ describe("Cloning", function() {
 
     network_one_id = 1000;
     network_two_id = 1001;
-    network_one = TestRPC.provider({network_id: network_one_id, seed: network_one_id});
-    network_two = TestRPC.provider({network_id: network_two_id, seed: network_two_id});
+    network_one = TestRPC.provider({network_id: network_one_id, seed: network_one_id, logger: log});
+    network_two = TestRPC.provider({network_id: network_two_id, seed: network_two_id, logger: log});
```
Comment (benjamincburns) [inline test/cloning.js]: Was this change necessary for it to run w/o error? I thought ganache-core's provider did this inside the constructor if the logger wasn't provided - would be sad if that's broken.

File: test/abstractions.js (hunk 2 above)
Comment (cgewecke) [reply test/abstractions.js]: Okay yes I agree, wasn't quite sure how to integrate this case into the rest of the tests.

File: contract.js [hunk 4]
```diff
@@ -172,8 +172,9 @@ var contract = (function(module) {
                   if (err) return reject(err);
 
                   // Reject on transaction failures, accept otherwise
+                  // Handles "0x00" or hex 0
                   if (receipt != null) {
-                    if (receipt.status === "0x00"){
+                    if (parseInt(receipt.status, 16) == 0){
```
Comment (benjamincburns) [inline contract.js]: Just an aside - it turns out that `parseInt` is happy w/o the base argument if there's a leading `0x` in the string. I'm not recommending that you change this again, though.

TIL :-)

File: test/cloning.js (hunk 3 above)
Comment (cgewecke) [reply test/cloning.js]: Right! It looks like the units are verbose in ganache-core [CI](https://travis-ci.org/trufflesuite/ganache-core/jobs/332481226#L2447)? 

Provider instantiated for that test [here](https://github.com/trufflesuite/ganache-core/blob/develop/test/mining.js#L13-L15)

Comment (benjamincburns) [reply test/cloning.js]: Hmm, perhaps I messed up the defaults, then - thanks for that!

File: test/util.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+var ganache = require("ganache-core");
+var Web3 = require("web3");
+
+var log = {
+  log: function(){}
+}
```
Comment (gnidan) [inline test/util.js]: Random idea: how do you feel about a strategy of using the npm [debug](https://www.npmjs.com/package/debug) package instead of just /dev/null'ing the logs entirely?

So you can see the logs if you want via something like `DEBUG=test:util:provider npm test`

Comment (cgewecke) [reply test/util.js]: I like that! This is a good time to add it :)

**Comment by tcoulter** [comment]
After talking to @cgewecke synchronously, we decided that when the `status` is `0`, `truffle-contract` should reject with an `Error` object with the following properties: 

* Useful error message (see below)
* `error.tx` - hash of transaction.
* `error.receipt` - full transaction receipt

The error message should include: 

* Something to the effect of "The transaction exited with an error (status = 0)"
* The transaction hash
* Recommendations / hints why this error occurred. e.g., OOG, `revert`, etc.

We also agreed that when Solidity eventually allows users to return their own error messages, that message should take precedence over this one (or be included in an umbrella message that tells the user a runtime error occurred, and "here's why...").

**Comment by cgewecke** [comment]
@gnidan This will need another approval sweep whenever it's considered for merging. Added a new `truffle-error`, some logic to discern between `invalid opcode` and `revert` error cases, and the message formatting Tim suggested above.

File: statuserror.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+var TruffleError = require("truffle-error");
+var inherits = require("util").inherits;
+var defaultGas = 90000;
```
Comment (gnidan) [inline statuserror.js]: Is there something special about this value that it's the default? Just curious.

Also, style pedantry: constant declarations should go after `require()` statements, mind moving this down 2 lines?

File: test/abstractions.js [hunk 7]
```diff
@@ -205,24 +204,83 @@ describe("Abstractions", function() {
...
+        return instance.runsOutOfGas();
+      }).then(function(){
+        assert.fail();
+      }).catch(function(e){
+        assert(e.message.includes('invalid opcode'));
+        assert(parseInt(e.receipt.status, 16) == 0)
+        done();
+      });
+    }).catch(done)
+  });
```
Comment (gnidan) [inline test/abstractions.js]: Love all the additional testing you've been adding!

File: statuserror.js (hunk 6 above)
Comment (cgewecke) [reply statuserror.js]: No idea why 90k is so magic. Hardcoded into the clients:
+ [ganache](https://github.com/trufflesuite/ganache-core/blob/8e080b4b2399f8504e5d66036f2951997980e1d5/lib/blockchain_double.js#L36)
+ [geth](https://github.com/ethereum/go-ethereum/blob/e61035c5a3630e4f6fd0fb3e5346a4eed8cedc80/internal/ethapi/api.go#L50)

Comment (benjamincburns) [reply statuserror.js]: @cgewecke 90k is part of the RPC spec (see `gas` param): https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sendtransaction

File: test/abstractions.js (hunk 7 above)
Comment (benjamincburns) [reply test/abstractions.js]: +1 to that!

**Comment by benjamincburns** [comment]
@tcoulter wrote:
>* **Recommendations / hints why this error occurred. e.g., OOG, revert, etc.**

We'll have enough info in the receipt to detect OOG, but perhaps we need to enlist `debug_traceTransaction` to fetch the actual runtime error? We'd just need to have a fallback mechanism when reaching out to clients which don't support it...

**Comment by tcoulter** [comment]
@benjamincburns I'd call that a separate ticket / feature. A nice to have, but would generally require significantly more engineering and so should be treated separately.

File: statuserror.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+var defaultGas = 90000;
+
+function StatusError(args, tx, receipt) {
+  var message;
+  var gasLimit = parseInt(args.gas) || defaultGas;
+
+  if(receipt.gasUsed === gasLimit){
+
+    message = "Transaction: " + tx + " exited with an error (status 0 - invalid opcode).\n" +
```
Comment (tcoulter) [inline statuserror.js]: I think `invalid opcode` here is a misnomer (i.e., the wrong error reason).

In the past, there was a time where Solidity produced `REVERT` opcodes before `ethereumjs-vm` knew how to understand them. This caused `ethereumjs-vm` to not know what to do when a `REVERT` was encountered, so it threw up its hands and said `invalid opcode`. This error message was always incorrect for the code that produced it, but it was something we had to live with until the vm caught up.

In this case, we specifically know the reason (and you check for it): Out of gas. I'd change `invalid opcode` here to `out of gas` to be more specific about the error reason.

In general, future `invalid opcode` errors mean something is wrong with the compiler or the vm.

Note: Correct me if there's another layer here where some opcode eats up all the gas. I thought they fixed all of those, but wouldn't be surprised if I'm wrong.

**Comment by tcoulter** [comment]
Also, in general, non-breakyness approved!

File: statuserror.js (hunk 8 above)
Comment (benjamincburns) [reply statuserror.js]: This error type should probably assert that `receipt.status` is `0`.

Also I agree that this code shouldn't try to appear as though it's passing an error from the EVM implementation if it's not actually doing so. If we're being pedantic, you can still get invalid opcode errors in esoteric cases, like bad copypasta'd bytecode, but they're far less likely to occur today than pre-byzantium. It could also be an invalid jump, in the case of bad inlined EVM assembler code.

File: statuserror.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+      "Please check that the transaction:\n" +
+      "    - satisfies all conditions set by Solidity `assert` statements.\n" +
```
Comment (benjamincburns) [inline statuserror.js]: also `require` statements.

Comment (benjamincburns) [reply statuserror.js]: Doh! My bad. I see you have that below.

File: statuserror.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      "    - has enough gas to execute all internal Solidity function calls.\n";
+
+  } else {
+
+    message = "Transaction: " + tx + " exited with an error (status 0 - revert).\n" +
+      "Please check that the transaction:\n" +
+      "    - satisfies all conditions set by Solidity `require` statements.\n" +
+      "    - does not trigger a Solidity `revert` statement.\n";
```
Comment (benjamincburns) [inline statuserror.js]: Fun one - might want to add `has enough gas to complete any expensive storage operations`. A rather confusing cause of reverts originates in the solidity runtime logic. That is, on certain operations it seems that the solidity compiler generates code which checks the gas limit for the transaction in place against a predicted gas cost for the next operation. If it determines that continuing with the expensive operation is going to blow the budget, it reverts rather than carrying on, saving you a bit of ETH, but wasting a the time that you'll spend scratching your head wondering how the hell it could be reverting *there* of all places.

Comment (benjamincburns) [reply statuserror.js]: I think fixed-size storage operations fit into this category, should you want to add this case to your tests (though they're already pretty kick ass as they are).

File: statuserror.js (hunk 8 above)
Comment (cgewecke) [reply statuserror.js]: @tcoulter @benjamincburns 

What do you think about revising along these lines?
```
Transaction: 0xabc..def exited with an error (status 0) after consuming all gas.
      Please check that the transaction:
         - satisfies all conditions set by Solidity `assert` statements.
         - has enough gas to execute all internal Solidity function calls.
         - does not trigger an invalid opcode by other means (ex: accessing an array out of bounds).
```

TBH trying to sneak 'invalid opcode' into this string since people are habituated to searching for it and this would be a cheap check for the assert case which fires `0xfe`, now canonized as the official invalid opcode with [EIP 141](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-141.md).

File: statuserror.js (hunk 10 above)
Comment (cgewecke) [reply statuserror.js]: @benjamincburns interesting!! Didn't know that. Will see if I can trigger it.

Comment (tcoulter) [reply statuserror.js]: You can trigger this by trying to create a contract (`CREATE` opcode) that doesn't fit into the gas limit. Solidity is saving you here by not consuming the gas required.

I'd lump all this in one message, i.e., turn this: 

> has enough gas to execute all internal Solidity function calls.

into

>  has enough gas to execute the full transaction.

File: statuserror.js (hunk 8 above)
Comment (benjamincburns) [reply statuserror.js]: > TBH trying to sneak 'invalid opcode' into this string since people are habituated to searching for it
Interesting observation. Prior to you saying this I would've encouraged you to avoid including the phrase altogether in an effort to try and break people's habits on that, but what you're suggesting is so much nicer. That is, I _really_ like the idea of actually using the old habit to guide people along, if not encourage new, better habits.

File: statuserror.js (hunk 10 above)
Comment (benjamincburns) [reply statuserror.js]: Agreed that my suggestion was too verbose. Separate idea for the future: we should link to docs pages tailored to these cases to give some more verbose explanation of what's going on.

File: statuserror.js (hunk 8 above)
Comment (tcoulter) [reply statuserror.js]: Ya, nice @cgewecke. Sounds good. 👍

File: statuserror.js (hunk 10 above)
Comment (gnidan) [reply statuserror.js]: @benjamincburns do you happen to know where you read that about Solidity runtime gas-cost anticipation? I'm curious because my first thought was "what about the halting problem?"

Comment (benjamincburns) [reply statuserror.js]: We spotted it in the bytecode one day early on when I was testing the byzantium changes. The halting problem doesn't apply because it's just being opportunistic and checking gas usage/limit immediately prior to an expensive operation against the calculable expected cost of that operation.

Comment (benjamincburns) [reply statuserror.js]: Also the halting problem doesn't apply to the EVM in general, specifically because gas limits are part of the machine definition.

Comment (benjamincburns) [reply statuserror.js]: @cgewecke & @tcoulter turns out you can see the various places where solidity inserts `REVERT`s by searching for 'appendConditionalRevert' [here](https://github.com/ethereum/solidity/blob/develop/libsolidity/codegen/ExpressionCompiler.cpp).
